zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/conditional/ConditionalBehavior.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.engine.processing.conditional;
+
+import io.camunda.zeebe.el.Expression;
+import io.camunda.zeebe.el.ExpressionLanguage;
+import io.camunda.zeebe.engine.processing.common.ExpressionProcessor;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.TypedCommandWriter;
+import io.camunda.zeebe.engine.state.immutable.ConditionalSubscriptionState;
+import io.camunda.zeebe.engine.state.immutable.ElementInstanceState;
+import io.camunda.zeebe.engine.state.immutable.ProcessingState;
+import io.camunda.zeebe.protocol.impl.record.value.conditional.ConditionalSubscriptionRecord;
+import io.camunda.zeebe.protocol.impl.record.value.variable.VariableRecord;
+import io.camunda.zeebe.protocol.record.intent.ConditionalSubscriptionIntent;
+import io.camunda.zeebe.protocol.record.intent.VariableIntent;
+import java.util.ArrayDeque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class ConditionalBehavior {
+
+  private final ElementInstanceState elementInstanceState;
+  private final ConditionalSubscriptionState conditionSubscriptionState;
+  private final TypedCommandWriter commandWriter;
+  private final ExpressionProcessor expressionProcessor;
+  private final ExpressionLanguage expressionLanguage;
+
+  public ConditionalBehavior(
+      final ProcessingState processingState,
+      final TypedCommandWriter commandWriter,
+      final ExpressionProcessor expressionProcessor,
+      final ExpressionLanguage expressionLanguage) {
+    elementInstanceState = processingState.getElementInstanceState();
+    conditionSubscriptionState = processingState.getConditionalSubscriptionState();
+    this.commandWriter = commandWriter;
+    this.expressionProcessor = expressionProcessor;
+    this.expressionLanguage = expressionLanguage;
+  }
+
+  /**
+   * Evaluates conditional subscriptions when variable events are present.
+   *
+   * <p>Key rules:
+   *
+   * <ol>
+   *   <li>Each subscription must be triggered at most once per document merge.
+   *   <li>For every scope in which a variable is created or updated, evaluate the conditional
+   *       subscriptions for that scope.
+   *   <li>Evaluation must proceed top-down. For each child scope, evaluate its conditionals as part
+   *       of the traversal.
+   * </ol>
+   *
+   * <p>Stop the traversal when either of the following occurs:
+   *
+   * <ul>
+   *   <li>An interrupting subscription is triggered.
+   *   <li>Subscription already triggered.
+   *   <li>No more child scopes exist.
```

Comment:
>Subscription already triggered.

❓ Wouldn't that mean that some subscriptions may not be triggered while they should be simply because another was already triggered?

I imagine that a higher-scoped subscription should be able to interrupt a lower-scoped one when both are triggered simultaneously.

EDIT: I see now that traversal is top-down. Perhaps this can be clarified at the top of the JavaDoc.